### ISSUE TITLE
Add bitwise operators?

### ISSUE THREAD

**Issue by eloff** [leafo/moonscript#91]
I ported some lua scripts recently, and because I do a lot of low-level stuff with the ffi I found that while overall the code become much more compact and readable, there was still the glaring ugliness of code like this:

import rshift, lshift, band, bor from bit
d0[j] = lshift(i, 2)
d1[j] = bor(rshift(band(i, 0x30), 4), lshift(band(i, 0x0f), 12))
d2[j] = bor(lshift(band(i, 0x03), 22), lshift(band(i, 0x3c), 6))
d3[j] = lshift(i, 16)

Which did not improve at all.

What do you think about adding bitwise operators? So that this code translates into the above?

d0[j] = i >> 2
d1[j] = ((i & 0x30) >>> 4) | ((i & 0x0f) << 12)
d2[j] =  ((i & 0x03) << 22) | ((i & 0x3c) << 6)
d3[j] = i << 16

Where the bit operators would be defined as for javascript (>> arithmetic right shift, >>> logical right shift) and would not be overloadable (if you're using them performance usually matters a lot!) Precedence would be as is universal in other languages.

The syntax is widely used in various languages and is quite concise. What do you think?

Clearly you'd have to have the bitops library installed for Lua or use LuaJIT for the compiled code to run.

**Comment by harraps** [comment]
I am digging out this old issue.
LUA 5.3 implements bitwise operators like
~ : NOT
& : AND
| : OR
~ : XOR
it seems that moonscript support them except for the xor
(a ~ b) get converted into (a (~b)) in LUA
either moonscript should have a xor operator like ~~
or (a ~ b) should be interpreted as (a ~ b)  and the use of parenthesis should be mandatory for expression like func(~b)

**Comment by kirbyfan64** [comment]
FWIW, this is going to get messy under LuaJIT...

**Comment by RyanSquared** [comment]
@kirbyfan64 - that's correct, but MoonScript shouldn't hinder the abilities of 5.3 because it might not be available in a different runtime. This is a bug across all Lua versions, not MoonScript. It's not up to MoonScript to deal with this incompatibility.

**Comment by Nymphium** [comment]
MoonScript already has the operators (https://github.com/leafo/moonscript/pull/209).

MoonScript is just a transcompiler and the runtime is on Lua (5.1, 5.2, ... and LuaJIT), so the language itself doesn't have to think about the compatibility.

**Comment by buckle2000** [comment]
@kirbyfan64  LuaJIT linked with Lua 5.2 support bitwise operators.

**Comment by nilnor** [comment]
@buckle2000 that's not true as far as I know (though I would love to be wrong), I'm guessing you thinking of something else. There's a open request for this [here](https://github.com/LuaJIT/LuaJIT/issues/63) fwiw.

**Comment by buckle2000** [comment]
@nilnor Yeah, you are right. I was misled by some non-official interpreter.

**Comment by RyanSquared** [comment]
@leafo mind closing this issue?

> MoonScript already has the operators (#209).

**Comment by harraps** [comment]
Really ?
Even with the last moonscript compiler v0.5.0, from this code:
`--moonscript`
`a = 5`
`b = 3`
`c = a ~ b`
`print(c)`
I get this output:
`--lua`
`local a = 5`
`local b = 3`
`local c = a(~b)`
`return print(c)`
To me, the issue is not fixed yet.

**Comment by RyanSquared** [comment]
That's an alternative issue then; the unary `~` takes precedence over the binary `~`.
